[clang-tidy] Add bugprone-smartptr-reset-ambiguous-call check

vbvictor opened the pull request:
Add new clang-tidy check that finds potentially erroneous calls to ``reset()`` method on smart pointers when
the pointee type also has a ``reset()`` method.

It's easy to make typo and delete object because the difference between ``.`` and ``->`` is really small.

Sometimes IDE's autocomplete will change ``->`` to ``.`` automatically. For example, developer wrote ``ptr->res`` but after pressing _Tab_ it became ``ptr.reset()``.

Small example:
```cpp
struct Resettable {
  void reset() { /* Own reset logic */ }
};

auto ptr = std::make_unique<Resettable>();

ptr->reset();  // Calls underlying reset method
ptr.reset();   // Makes the pointer null

// After Fix-its
(*ptr).reset();  // Clearly calls underlying reset method
ptr = nullptr;   // Clearly makes the pointer null
```

Fixes #120908

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/reset-call.rst:
```diff
@@ -0,0 +1,33 @@
+bugprone-reset-call
+===================
+
+Finds calls to ``reset()`` method on smart pointers where the pointee type
```
Please make this statement same as statement in Release Notes.

vbvictor replied:
Done, also removed parentheses in ``reset`` to follow general style

PiotrZSL reviewed (changes requested):
1. Consider TK_IgnoreUnlessSpelledInSource, may simplify matchers
2. Name is too generic, consider:
 - bugprone-smartptr-reset-pointee-reset
 - bugprone-smartptr-reset-call
 - bugprone-smartptr-pointee-reset
 - bugprone-smartptr-reset-ambiguous-call (my prefered)

Or any simillar.

vbvictor commented:
> 1. Consider TK_IgnoreUnlessSpelledInSource, may simplify matchers
> 2. Name is too generic, consider:
> 
> * bugprone-smartptr-reset-pointee-reset
> * bugprone-smartptr-reset-call
> * bugprone-smartptr-pointee-reset
> * bugprone-smartptr-reset-ambiguous-call (my prefered)
> 
> Or any simillar.

1. Could not find a way to use TK_IgnoreUnlessSpelledInSource because code in function
`template <typename T>
void TemplatePositiveTest() {}`
is implicitly instantiated for different template types, so it is being ignored by matchers with TK_IgnoreUnlessSpelledInSource.
2. Renamed check to bugprone-smartptr-reset-ambiguous-call.

5chmidti commented on clang-tools-extra/clang-tidy/bugprone/SmartptrResetAmbiguousCallCheck.cpp:
```diff
@@ -0,0 +1,134 @@
+
+void SmartptrResetAmbiguousCallCheck::check(
+    const MatchFinder::MatchResult &Result) {
+  const auto *SmartptrResetCall =
+      Result.Nodes.getNodeAs<CXXMemberCallExpr>("smartptrResetCall");
+  const auto *ObjectResetCall =
+      Result.Nodes.getNodeAs<CXXMemberCallExpr>("objectResetCall");
```
Please move these into each of the `if` conditions to potentially save the one lookup from `objectResetCall`:

```c++
if (const auto *SmartptrResetCall = Result.Nodes.getNodeAs<CXXMemberCallExpr>("smartptrResetCall")) {
...
}
```

vbvictor replied:
Done

5chmidti commented on clang-tools-extra/clang-tidy/bugprone/SmartptrResetAmbiguousCallCheck.cpp:
```diff
@@ -0,0 +1,134 @@
+    DiagBuilder << FixItHint::CreateReplacement(
+        SourceRange(Member->getMemberLoc(), SmartptrResetCall->getEndLoc()),
+        " nullptr");
+  }
```
At this point you can return for inside this `if` branch, because the other case will never happen in the same `check` call

vbvictor replied:
Done

5chmidti reviewed (changes requested):
I think that this check is one that we don't want to emit fixes for, at least not in the main diagnostic, but maybe as a fixit attached to a note. Sort of like: 
```
warning: ...
note: (be more explicit and) use '= nullptr'
```

The diagnostic message is also too generic and does not say what the problem is. One potential wording could be: `be explicit when calling 'reset()' on a smart-pointer with a pointee that has a 'reset()' method`

What do others think?

denzor200 commented on clang-tools-extra/clang-tidy/bugprone/SmartptrResetAmbiguousCallCheck.h:
```diff
@@ -0,0 +1,34 @@
+  void registerMatchers(ast_matchers::MatchFinder *Finder) override;
+  void check(const ast_matchers::MatchFinder::MatchResult &Result) override;
+  bool isLanguageVersionSupported(const LangOptions &LangOpts) const override {
+    return LangOpts.CPlusPlus;
```
I disagree, the code issue this check is hitting for also relevant for C++03. Think about `boost::shared_ptr`, for example.

vbvictor replied:
As for now, this check only matches `::std::unique_ptr` and `::std::shared_ptr`, so this comment is valid.
In the future, I think a list of smartptr-like classes should be an option for check.

vbvictor replied:
I added an option to specify custom smart pointers. Now `boost::shared_ptr` can be valid pointer for this check, so I did not change to `CPlusPlus11`.

PiotrZSL replied:
Ok, but then add "boost::shared_ptr" to default configuration

EugeneZelenko commented on clang-tools-extra/clang-tidy/bugprone/SmartptrResetAmbiguousCallCheck.cpp:
```diff
@@ -0,0 +1,134 @@
+  if (ObjectResetCall) {
+    const auto *Arrow = Result.Nodes.getNodeAs<CXXOperatorCallExpr>("OpCall");
+
+    const auto SmartptrSourceRange =
```
`auto` should not be used here - type is not spelled explicitly.

vbvictor replied:
Done

vbvictor commented:
> I think that this check is one that we don't want to emit fixes for, at least not in the main diagnostic, but maybe as a fixit attached to a note

> What do others think?

vbvictor commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/smartptr-reset-ambiguous-call.rst:
```diff
@@ -0,0 +1,34 @@
+rather than resetting its state.
+It's easy to make such typo because the difference between ``.`` and ``->`` is really small.
+
+The recommended approach is to make the intent explicit by using either member access or direct assignment:
```
Done

vbvictor commented on clang-tools-extra/clang-tidy/bugprone/BugproneTidyModule.cpp:
```diff
@@ -144,6 +145,8 @@ class BugproneModule : public ClangTidyModule {
     CheckFactories.registerCheck<IncorrectEnableIfCheck>(
         "bugprone-incorrect-enable-if");
+    CheckFactories.registerCheck<SmartptrResetAmbiguousCallCheck>(
+        "bugprone-smartptr-reset-ambiguous-call");
```
Done

vbvictor commented:
> I think that this check is one that we don't want to emit fixes for, at least not in the main diagnostic, but maybe as a fixit attached to a note. 

I think the case when `reset` is called on smart pointer is trivial for a fixit, so it can be in the main diagnostic.
```cpp
ptr.reset();
// Converts to
ptr = nullptr;
```
A `reset` call on an underlying object can be tricky to fix, so a note with fix is a great idea.

> The diagnostic message is also too generic and does not say what the problem is. One potential wording could be: `be explicit when calling 'reset()' on a smart-pointer with a pointee that has a 'reset()' method`

I agree that my wording is a bit confusing, so I will change it to a better one when fixit format is settled.

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/smartptr-reset-ambiguous-call.rst:
```diff
@@ -0,0 +1,47 @@
+
+.. option:: SmartPointers
+
+    Semicolon-separated list of class names of custom smart pointers.
```
Please add default value.

vbvictor replied:
Done

vbvictor commented:
Thank you everyone for the feedback!

@PiotrZSL I tried using TK_IgnoreUnlessSpelledInSource with `cxxDependentScopeMemberExpr` ast-matcher but failed to write a working matcher case with nested template parameters:
```cpp

template <typename T>
void TemplatePositiveTest() {
  std::unique_ptr<T> u_ptr;

  u_ptr.reset();
  u_ptr->reset();
}

void instantiate() {
  TemplatePositiveTest<std::unique_ptr<int>>();
}
```

As for now, I'm thinking about leaving it as is and make an NFC change in the future when I get more familiar with matchers.
Apart from TK_IgnoreUnlessSpelledInSource I fixed your comments.

@5chmidti I fixed all your comments and added better _warning_ and _note_ messages:

```cpp
  s.reset(); // warning: be explicit when calling 'reset()' on a smart pointer with a pointee that has a 'reset()' method
             // note: assign the pointer to 'nullptr'
  s->reset(); // warning: be explicit when calling 'reset()' on a pointee of a smart pointer
              // note: use dereference to call 'reset' method of the pointee
```

@EugeneZelenko Fixed all pr comments.

HerrCai0907 commented on clang-tools-extra/clang-tidy/bugprone/SmartptrResetAmbiguousCallCheck.cpp:
```diff
@@ -0,0 +1,150 @@
+
+AST_MATCHER_P(CallExpr, everyArgumentMatches,
+              ast_matchers::internal::Matcher<Expr>, InnerMatcher) {
+  if (Node.getNumArgs() == 0)
+    return true;
+
```
looks like useless since it is return at the end if it is empty.

vbvictor replied:
Yes, removed useless condition.

HerrCai0907 commented on clang-tools-extra/clang-tidy/bugprone/SmartptrResetAmbiguousCallCheck.cpp:
```diff
@@ -0,0 +1,150 @@
+}
+
+AST_MATCHER(CXXMethodDecl, hasOnlyDefaultParameters) {
+  if (Node.param_empty())
```
same

vbvictor replied:
Yes, removed useless condition.

HerrCai0907 reviewed (approved):
LGTM.

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/smartptr-reset-ambiguous-call.rst:
```diff
@@ -0,0 +1,48 @@
+bugprone-smartptr-reset-ambiguous-call
+======================================
+
+Finds potentially erroneous calls to ``reset`` method on
```
Please fill as much as possible to 80-characters limit.

vbvictor replied:
Done

vbvictor commented:
Ping

EugeneZelenko commented:
Please rebase from `main`. Release Notes were cleared after 20 branch creation.

vbvictor commented:
Ping

vbvictor commented:
@PiotrZSL, @5chmidti Ping

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/SmartptrResetAmbiguousCallCheck.cpp:
```diff
@@ -0,0 +1,146 @@
+                 recordType(hasDeclaration(TypeWithReset)))))));
+
+  // Find a.reset() calls
+  Finder->addMatcher(
+      cxxMemberCallExpr(callee(memberExpr(member(hasName("reset")))),
+                        everyArgumentMatches(cxxDefaultArgExpr()),
+                        on(expr(hasType(SmartptrWithBugproneReset))))
+          .bind("smartptrResetCall"),
+      this);
+
+  // Find a->reset() calls
+  Finder->addMatcher(
```
you could merge both addMatcher by using "anyOf".

vbvictor replied:
I Could not find a good way to use "anyOf" in this context.
If I write like this, it gives CE.
```cpp
Finder->addMatcher(anyOf(...));
```
If I write:
```cpp
Finder->addMatcher(
    cxxMemberCallExpr(
       anyOf(...))
    .bind("MemberCall"));
```
I get only one bind-name for `cxxMemberCallExpr` which complicates futher processing.
If there aren't any performance issues, I'd like to leave as is.

PiotrZSL replied:
maybe something like this: 
```
Finder->addMatcher(
      cxxMemberCallExpr(
      	   callee(ResetMethod),
           unless(hasAnyArgument(expr(unless(cxxDefaultArgExpr())))),
           anyOf(
                 thisPointerType(SmartptrWithBugproneReset),
                 on(cxxOperatorCallExpr(
                      hasOverloadedOperatorName("->"),
                      hasArgument(
                          0, expr(hasType(classTemplateSpecializationDecl(IsSmartptr)))))
                   )
           )
       ), this);
```

vbvictor replied:
With some modifications made this work,
changed `thisPointerType()` to `on(expr())` because it failed to match some cases with nested ptr's.
`std::unique_ptr<std::shared_ptr<std::unique_ptr<Resettable>>> nested_ptr;`

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/SmartptrResetAmbiguousCallCheck.cpp:
```diff
@@ -0,0 +1,146 @@
+    const auto *Member = cast<MemberExpr>(SmartptrResetCall->getCallee());
+
+    diag(SmartptrResetCall->getBeginLoc(),
+         "be explicit when calling 'reset()' on a smart pointer with a "
+         "pointee that has a 'reset()' method");
```
consider format "whats wrong, how to fix". In this case warning probably could start with `ambiguous call to `reset()` on a smart pointer with pointee that also has a 'reset()' method, prefer more explicit approach`

vbvictor replied:
Done

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/SmartptrResetAmbiguousCallCheck.cpp:
```diff
@@ -0,0 +1,146 @@
+         "be explicit when calling 'reset()' on a smart pointer with a "
+         "pointee that has a 'reset()' method");
+
+    diag(SmartptrResetCall->getBeginLoc(), "assign the pointer to 'nullptr'",
```
again, `consider assigning 'nullptr' here`

vbvictor replied:
Done

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/SmartptrResetAmbiguousCallCheck.cpp:
```diff
@@ -0,0 +1,146 @@
+                              *Result.SourceManager, getLangOpts());
+
+    diag(ObjectResetCall->getBeginLoc(),
+         "be explicit when calling 'reset()' on a pointee of a smart pointer");
```
same comments related to msg as before

vbvictor replied:
Done

PiotrZSL commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/smartptr-reset-ambiguous-call.rst:
```diff
@@ -0,0 +1,47 @@
+
+.. option:: SmartPointers
+
+    Semicolon-separated list of class names of custom smart pointers.
```
add fully qualified

vbvictor replied:
Done

PiotrZSL reviewed (commented):
1. Overall ok, please work a little bit on an warning message.
2. Probably better name would be: readability-ambiguous-smartptr-reset-call
3. Same issue exist on other objects like std::optional, and other methods like std::optional<std::set> and call to `emplace`. Maybe more checks or more generic check could be constructed where configuration is an pair of "object" and "method".

Consider points 1 & 2. And after that it could be delivered.

vbvictor commented:
@PiotrZSL, thank you for pointing out better naming pattern for diag-messages, would use it in the future.

1. Created new warning and fix-it messages.
2. Renamed check.
3. As for now, my check can support any type that will have `reset` method, this include `std::optional`. Maybe a more general approach would be to detect overlapping methods with classes that are returned via `operator*` and `operator->`. However, if some method is called on main object and there is no equivalent of `= nullptr` I don't know how user can fix such warning.

PiotrZSL commented on clang-tools-extra/docs/clang-tidy/checks/readability/ambiguous-smartptr-reset-call.rst:
```diff
@@ -0,0 +1,48 @@
+.. option:: SmartPointers
+
+    Semicolon-separated list of fully qualified class names of custom smart
+    pointers. Default value is `::std::unique_ptr;::std::shared_ptr;
```
out of sync, change also DefaultSmartPointers to match this, consider adding std::optional here also (even that is not smart ptr)

vbvictor replied:
Synced code and docs, sorry for inconvenience.
Also added tests to cover all listed types.

denzor200 replied:
> consider adding std::optional here also (even that is not smart ptr)

Strictly disagree with you, optional-like classes are not assignable with `nullptr`.
Here is no place for them in this check otherwise it will produce wrong fix hints:

```
{
  std::optional<Resettable> o;
  o.reset(); // should produce `o = std::nullopt;` but `o = nullptr;` does
}
{
  boost::optional<Resettable> o;
  o.reset(); // should produce `o = boost::none;` but `o = nullptr;` does
}

```

https://godbolt.org/z/YM79G8zdP

vbvictor replied:
Since this check only emit fixes with `--fix-notes`, there will be no mistakes in a normal clang-tidy run.

PiotrZSL commented on clang-tools-extra/docs/clang-tidy/checks/readability/ambiguous-smartptr-reset-call.rst:
```diff
@@ -0,0 +1,48 @@
+  ptr = nullptr;   // Clearly makes the pointer null
+
+The default smart pointers that are considered are ``std::unique_ptr``,
+``std::shared_ptr``, ``boost::unique_ptr``, ``boost::shared_ptr``. To specify
```
boost does not have unique_ptr, it has boost::scoped_ptr

vbvictor replied:
Fixed

PiotrZSL commented on clang-tools-extra/test/clang-tidy/checkers/readability/ambiguous-smartptr-reset-call.cpp:
```diff
@@ -0,0 +1,239 @@
+
+    ptr.reset();
+  }
+
```
add tests where smart ptr is behind typedef like:
`using SmartPtr = std::unique_ptr<Resettable>;`
Simply because there is chance that it may not be detected.

denzor200 commented on clang-tools-extra/clang-tidy/readability/AmbiguousSmartptrResetCallCheck.cpp:
```diff
@@ -0,0 +1,131 @@
+}
+
+const auto DefaultSmartPointers =
+    "::std::shared_ptr;::std::unique_ptr;::std::optional;"
```
remove `std::optional` from this check

denzor200 commented on clang-tools-extra/docs/clang-tidy/checks/readability/ambiguous-smartptr-reset-call.rst:
```diff
@@ -0,0 +1,49 @@
+
+    Semicolon-separated list of fully qualified class names of custom smart
+    pointers. Default value is `::std::unique_ptr;::std::shared_ptr;
+    ::std::optional;::boost::unique_ptr;::boost::shared_ptr`.
```
`::boost::unique_ptr` -> `::boost::scoped_ptr`

vbvictor replied:
Done

denzor200 commented on clang-tools-extra/clang-tidy/readability/AmbiguousSmartptrResetCallCheck.cpp:
```diff
@@ -0,0 +1,130 @@
+}
+
+const auto DefaultSmartPointers = "::std::shared_ptr;::std::unique_ptr;"
+                                  "::boost::shared_ptr;::boost::scoped_ptr";
```
yet another one wrong fix hint, `::boost::scoped_ptr` doesn't have `operator=` at all :(
No way to reset this pointer without `reset`, except the crutch like calling `swap` method with a reference to default initialized object passed.

denzor200 replied:
Since this check is unable to handle the issue with `::boost::scoped_ptr`, here I requested another check to get rid of scoped_ptr: https://github.com/llvm/llvm-project/issues/128179

vbvictor replied:
Thank you for pointing that out. I have written a note in documentation that says the check may not provide correct fix-its on user-defined types. I guess I will also include `::boost::scoped_ptr` as an example.
As for `::boost::scoped_ptr` the correct fix-it would be `ptr.reset(nullptr)` or `ptr.reset(0)`

vbvictor replied:
I think we should stick to 100% correct fix-its for default-defined types and warn users about incorrect behavoir of other types that lack `operator=` or `operator=` doesn't expect `std::nullptr_t`.

vbvictor commented:
@PiotrZSL, @denzor200 
I think we should provide 100% valid fix-its for default `SmartPointers`. That will be `std::shared_ptr`, `std::unique_ptr`, `boost::shared_ptr`. I created a note in docs that warns users about inconvenience that can happen when non-default smart pointers are specified.
May you please review one more time.
If everything is good i'd need help in merging this pr since i don't have write access to main

PiotrZSL commented on clang-tools-extra/clang-tidy/readability/AmbiguousSmartptrResetCallCheck.cpp:
```diff
@@ -0,0 +1,129 @@
+
+  if (const auto *Arrow =
+          Result.Nodes.getNodeAs<CXXOperatorCallExpr>("ArrowOp")) {
+    const auto *ObjectResetCall =
+        Result.Nodes.getNodeAs<CXXMemberCallExpr>("MemberCall");
```
duplicated member call, just move it before if and use in both if's

PiotrZSL commented on clang-tools-extra/clang-tidy/readability/AmbiguousSmartptrResetCallCheck.cpp:
```diff
@@ -0,0 +1,129 @@
+        Result.Nodes.getNodeAs<CXXMemberCallExpr>("MemberCall");
+
+    const CharSourceRange SmartptrSourceRange =
+        Lexer::getAsCharRange(Arrow->getArg(0)->getSourceRange(),
```
maybe just 'bind' an object to some variable and read it like an MemberCall

vbvictor replied:
Binding to variable caused test-failures, and since it isn't a "hot path" I'd prefer to leave as is.

PiotrZSL commented on clang-tools-extra/docs/clang-tidy/checks/readability/ambiguous-smartptr-reset-call.rst:
```diff
@@ -0,0 +1,57 @@
+
+    Semicolon-separated list of fully qualified class names of custom smart
+    pointers. Default value is `::std::unique_ptr;::std::shared_ptr;
+    ::boost::shared_ptr`.
```
```suggestion
    ::boost::scoped_ptr`.
```

vbvictor replied:
I think here my code is correct since `::boost::scoped_ptr` does not have `operator=` which causes fix-its to be invalid.

PiotrZSL replied:
That's not about fixes, it is about default settings for configuration option.

vbvictor replied:
As I said in https://github.com/llvm/llvm-project/pull/121291#issuecomment-2676086944, I think it's better to provide default pointers whose fixes will be 100% correct (hopefully).
You think we should put values that provide incorrect fix-its sometimes?

PiotrZSL commented on clang-tools-extra/test/clang-tidy/checkers/readability/ambiguous-smartptr-reset-call-custom-pointers.cpp:
```diff
@@ -0,0 +1,68 @@
+
+namespace std {
+
+template <typename T>
```
consider including clang-tools-extra/test/clang-tidy/checkers/modernize/Inputs/smart-ptr/shared_ptr.h and clang-tools-extra/test/clang-tidy/checkers/modernize/Inputs/smart-ptr/unique_ptr.h instead

PiotrZSL commented on clang-tools-extra/docs/clang-tidy/checks/readability/ambiguous-smartptr-reset-call.rst:
```diff
@@ -0,0 +1,57 @@
+``std::unique_ptr``, ``std::shared_ptr``, ``boost::shared_ptr``. To specify
+other smart pointers or other classes use the :option:`SmartPointers` option.
+
+
```
Mention in documentation that fixes are enabled via `--fix-notes`

PiotrZSL commented on clang-tools-extra/test/clang-tidy/checkers/readability/ambiguous-smartptr-reset-call-custom-pointers.cpp:
```diff
@@ -0,0 +1,68 @@
+// RUN: %check_clang_tidy %s readability-ambiguous-smartptr-reset-call %t \
```
you could merge this file into main test, just have 2 RUN commands there and specify custom suffix for this one (you may need to include current default values also)

vbvictor replied:
I'd prefer to leave two separate files, they are easier to read and maintain imho.

PiotrZSL reviewed (approved):
In general looks fine.
Few nits.

vbvictor committed:
add ambigous-smartptr-reset-call check

vbvictor commented:
Since all approved are made, can we land this PR?
@HerrCai0907

vbvictor commented:
Ping @HerrCai0907, could you please address https://github.com/llvm/llvm-project/pull/121291#issuecomment-2702153354. Thank you!
Can I ask someone to grant me write access to repository?
Since you basically the only one who merges PRs of other people, this could lower amount of work you need to do.
Also, I'd be happy to help new contributors in merging their PRs.

EugeneZelenko commented:
> @HerrCai0907, could you please address [#121291 (comment)](https://github.com/llvm/llvm-project/pull/121291#issuecomment-2702153354). Thank you! Can I ask someone to grant me write access to repository? Since you basically the only one who merges PRs of other people, this could lower amount of work you need to do. Also, I'd be happy to help new contributors in merging their PRs.

If you have at least three merged pull requests, you could request commit access using [other ones](https://github.com/llvm/llvm-project/issues/?q=is%3Aissue%20state%3Aopen%20%20label%3Ainfra%3Acommit-access-request) as example.

PiotrZSL commented:
@vbvictor If you don't want this to be merged as "70346889+vbvictor@users.noreply.github.com.", please change your email privacy settings.

vbvictor commented:
> @vbvictor If you don't want this to be merged as '[70346889+vbvictor@users.noreply.github.com](mailto:70346889+vbvictor@users.noreply.github.com)', please change your email privacy settings.

Thank you for noticing!
I changed my email to public visibility.
